lib/spack/spack/util/compression.py

```diff
@@ -23,37 +26,222 @@ def allowed_archive(path):
+        f_bz = bz2.BZ2File(archive_file, mode='rb')
+        with open(archive_out, 'wb') as ar:
+            ar.write(f_bz.read())
+        f_bz.close()
+    except ImportError:
+        bunzip2 = which('bunzip2', required=True)
+        bunzip2.add_default_arg('-q')
+        return bunzip2(archive_file)
+    return archive_out
+
+
 def _gunzip(archive_file):
-    """Like gunzip, but extracts in the current working directory
+    """ Decompress `.gz` extensions. Prefer native Python `gzip` module.
+    Failing back to system utility gunzip.
+    Like gunzip, but extracts in the current working directory
     instead of in-place.
 
     Args:
         archive_file (str): absolute path of the file to be decompressed
     """
-    import gzip
-    decompressed_file = os.path.basename(archive_file.strip('.gz'))
+    _, ext = os.path.splitext(archive_file)
+    decompressed_file = os.path.basename(archive_file.strip(ext))
     working_dir = os.getcwd()
     destination_abspath = os.path.join(working_dir, decompressed_file)
-    with gzip.open(archive_file, "rb") as f_in:
+    try:
+        import gzip
+        f_in = gzip.open(archive_file, "rb")
         with open(destination_abspath, "wb") as f_out:
             f_out.write(f_in.read())
+    except ImportError:
+        gzip = which("gzip")
+        gzip.add_default_arg("-d")
+        gzip(archive_file)
+    return destination_abspath
+
 
+def _unzip(archive_file):
+    """Try to use Python's zipfile, but extract in the current working
+    directory instead of in-place.
 
-def decompressor_for(path, extension=None):
+    If unavailable, try system unzip.
+
+    Args:
+        archive_file (str): absolute path of the file to be decompressed
+    """
+    destination_abspath = os.getcwd()
+    try:
+        from zipfile import ZipFile
+        zf = ZipFile(archive_file, 'r')
+        zf.extractall(destination_abspath)
+        zf.close()
+    except ImportError:
+        unzip = which('unzip', required=True)
+        unzip.add_default_arg('-q')
+        unzip(archive_file)
+    return destination_abspath
+
+
+def composer(funcA):
+    """Utility method currying function pointers
+    returns a function pointer to be called with
+    a function to be curried with the current function
+    argument.
+
+    Args:
+        funcA : Function to be curried.
+    """
+    def b(funcB):
+        def c(*args, **kwargs):
+            return funcA(funcB(*args, **kwargs))
+        return c
+    return b
+
+
+def decompressor_for(path, ext=None):
+    """Wrapper for select_decompressor_for, returns
+    a function pointer to appropriate decompression
+    algorithm.
+
+    On Unix - simply invokes select_decompressor_for
+
+    On Windows - archives with one extension are
+    passed through to select_decompressor_for. Multiple
+    extension archives are decomposed into their component
+    extensions and the requsite decompression algorithms are
+    curried and returned as a single callable function pointer.
```

Comment:
> I think I see what you're getting at but (and maybe I'm reducing the problem a little too far here) that essentially sounds like what we're doing now with a more explicit understanding that we will not handle more than two file exts (which I believe we can actually support, I just wasn't sure if the case came up in Spack).

My current understanding/opinions are:

* (assumption) This solution in this PR doesn't currently handle > 2 extensions (but it could easily be extended to)
* The solution is rather complex IMO in terms of handling the case of 2 extensions for Windows, so I'd prefer to see an object explicitly created for handling 2 extensions
* (assumption) I'm not aware of compression/archiving combos leading to > 2 extensions

If either of my assumptions are false, then I would revisit the value of this code. IMO it's a bit complicating to ad-hoc introduce a generic currying mechanism into the compression code though.

Reply:
My initial thought was to make it easily extensible to handle an arbitrary number of extensions but to explicitly only handle two for the time being as I also am not aware of a case with >2 extensions. You are correct that this code will ignore any more than two extensions. The ad-hoc currying is potentially pretty fragile, but I thought it would be a little easier to extend down the line (again, I can't imagine a reason for having three file extensions in this context), but I can refactor to an object no problem.